mono/metadata/threads.c

```diff
@@ -464,35 +502,133 @@ static void ensure_synch_cs_set (MonoInternalThread *thread)
 	synch_cs = g_new0 (MonoCoopMutex, 1);
 	mono_coop_mutex_init_recursive (synch_cs);
 
-	if (mono_atomic_cas_ptr ((gpointer *)&thread->synch_cs,
-					       synch_cs, NULL) != NULL) {
+	gpointer ret = mono_atomic_cas_ptr ((gpointer *)&thread->synch_cs, synch_cs, NULL);
+
+	if (ret != NULL) {
 		/* Another thread must have installed this CS */
+		/* Or the finalizer had to have run */
 		mono_coop_mutex_destroy (synch_cs);
 		g_free (synch_cs);
 	}
 }
 
-static inline void
-lock_thread (MonoInternalThread *thread)
+static void
+internal_thread_free (MonoInternalThread *thread)
+{
+	MonoCoopMutex *synch_cs = thread->synch_cs;
+	thread->synch_cs = INTERNAL_THREAD_TOMBSTONE;
+	mono_memory_barrier ();
+
+	mono_coop_mutex_destroy (synch_cs);
+	g_free (synch_cs);
+}
+
+static void
+cleanup_thread_lock (MonoInternalThread *thread)
 {
-	if (!thread->synch_cs)
+	mono_memory_barrier ();
+	MonoInternalThreadLockFlags flags;
+	flags.placeholder = mono_atomic_load_i64 (&thread->lock_flags.placeholder);
+
+	while (!flags.freeing) {
+		MonoInternalThreadLockFlags old_flags = flags;
+		flags.freeing = TRUE;
+		flags.placeholder = mono_atomic_cas_i64 (&thread->lock_flags.placeholder, flags.placeholder, old_flags.placeholder);
+	}
+	mono_memory_barrier ();
+
+	// If we started cleanup without any writers, finish the freeing
+	// ourselves. Else the thread who unlocks the thread after the last pending 
+	// wait is finished will do the freeing.
+	if (flags.waiters == 0)
+		internal_thread_free (thread);
+}
+
+static void
+change_wait_queue (MonoInternalThread *thread, int diff)
+{
+	mono_memory_barrier ();
+	MonoInternalThreadLockFlags flags;
+	flags.placeholder = mono_atomic_load_i64 (&thread->lock_flags.placeholder);
+
+	while (TRUE) {
+		MonoInternalThreadLockFlags old_flags = flags;
+		flags.waiters += diff;
+		MonoInternalThreadLockFlags seen_old_flags;
+		seen_old_flags.placeholder = mono_atomic_cas_i64 (&thread->lock_flags.placeholder, flags.placeholder, old_flags.placeholder);
+
+		// flags was set.
+		if (seen_old_flags.placeholder == old_flags.placeholder)
+			break;
+		else
+			flags.placeholder = seen_old_flags.placeholder;
+	}
+
+	if (flags.waiters == 0 && flags.freeing)
+		internal_thread_free (thread);
+	else if (!thread->synch_cs)
 		ensure_synch_cs_set (thread);
+}
```

Comment:
You should maybe assert that diff is +1 or -1 and the logic at the end should vary based on it.
Only check waiting/freeing if diff==-1, only check sync_cs if diff==+1.
It is probably correct asis though.

This still seems off. In my proposal I decremened and set free at the same time.
I did it inside the lock, then unlocked, then freed the lock.
Perhaps other orders are ok.
I only had one call to internal_thread_free though which seems better.

I would name `placeholder` as `i` or `integer` or `atomic_integer` perhaps.
People often calls AsUint32 or AsSizeT -- i.e. its type. It is the "union views as typeX".

Reply:
If you decrement before you unlock the lock then can't someone else free the lock before we unlock?

Reply:
Ok that last comment wasn't entirely  accurate.
It looks like I split setting free and decrement.
Still the structure here seems a little different.